lib/compat/wordpress-6.8/blocks.php

```diff
@@ -144,3 +144,165 @@ function gutenberg_stabilize_experimental_block_supports( $args ) {
+
+function gutenberg_apply_block_hooks_to_post_content( $content ) {
+	// The `the_content` filter does not provide the post that the content is coming from.
+	// However, we can infer it by calling `get_post()`, which will return the current post
+	// if no post ID is provided.
+	return apply_block_hooks_to_content( $content, get_post(), 'insert_hooked_blocks' );
+}
+// We need to apply this filter before `do_blocks` (which is hooked to `the_content` at priority 9).
+add_filter( 'the_content', 'gutenberg_apply_block_hooks_to_post_content', 8 );
+
+/**
+ * Hooks into the REST API response for the Posts endpoint and adds the first and last inner blocks.
+ *
+ * @since 6.6.0
+ * @since 6.8.0 Support non-`wp_navigation` post types.
+ *
+ * @param WP_REST_Response $response The response object.
+ * @param WP_Post          $post     Post object.
+ * @return WP_REST_Response The response object.
+ */
+function gutenberg_insert_hooked_blocks_into_rest_response( $response, $post ) {
+	if ( empty( $response->data['content']['raw'] ) || empty( $response->data['content']['rendered'] ) ) {
+		return $response;
+	}
+
+	$attributes            = array();
+	$ignored_hooked_blocks = get_post_meta( $post->ID, '_wp_ignored_hooked_blocks', true );
+	if ( ! empty( $ignored_hooked_blocks ) ) {
+		$ignored_hooked_blocks  = json_decode( $ignored_hooked_blocks, true );
+		$attributes['metadata'] = array(
+			'ignoredHookedBlocks' => $ignored_hooked_blocks,
+		);
+	}
+
+	if ( 'wp_navigation' === $post->post_type ) {
+		$wrapper_block_type = 'core/navigation';
+	} else {
+		$wrapper_block_type = 'core/post-content';
+	}
+
+	$content = get_comment_delimited_block_content(
+		$wrapper_block_type,
+		$attributes,
+		$response->data['content']['raw']
+	);
+
+	$content = apply_block_hooks_to_content(
+		$content,
+		$post,
+		'insert_hooked_blocks_and_set_ignored_hooked_blocks_metadata'
+	);
+
+	// Remove mock block wrapper.
+	$content = remove_serialized_parent_block( $content );
+
+	$response->data['content']['raw'] = $content;
+
+	// No need to inject hooked blocks twice.
+	$priority = has_filter( 'the_content', 'apply_block_hooks_to_content' );
+	if ( false !== $priority ) {
+		remove_filter( 'the_content', 'apply_block_hooks_to_content', $priority );
+	}
+
+	/** This filter is documented in wp-includes/post-template.php */
+	$response->data['content']['rendered'] = apply_filters( 'the_content', $content );
+
+	// Add back the filter.
+	if ( false !== $priority ) {
+		add_filter( 'the_content', 'apply_block_hooks_to_content', $priority );
+	}
+
+	return $response;
+}
+add_filter( 'rest_prepare_page', 'gutenberg_insert_hooked_blocks_into_rest_response', 10, 2 );
+add_filter( 'rest_prepare_post', 'gutenberg_insert_hooked_blocks_into_rest_response', 10, 2 );
+
+/**
+ * Updates the wp_postmeta with the list of ignored hooked blocks
+ * where the inner blocks are stored as post content.
+ *
+ * @since 6.6.0
+ * @since 6.8.0 Support non-`wp_navigation` post types.
```

Comment:
How about being more explicit? 
```suggestion
 * @since 6.6.0 Support `wp_navigation` post type.
 * @since 6.8.0 Support `post` and `page` post types.
```

Reply:
I wanted to keep the filter as generic as possible, in order to potentially allow the user to potentially attach it to other post types' hooks.

FWIW, for the Core version of this change, I've actually been considering if we should apply Block Hooks directly from inside the Posts controller, rather than adding the filter to `rest_prepare_page` and `rest_prepare_post`, see https://github.com/WordPress/wordpress-develop/pull/7898#discussion_r1867470796:

> This [currently] follows the precedent of `rest_prepare_wp_navigation` below. However, it has the unfortunate side effect that Block Hooks aren't automatically applied to user-defined CPTs; those would need to be added manually (via `add_filter( 'rest_prepare_my_cpt', ... )`). (AFAIK, there's no "generic" `rest_prepare` filter that could cover all post types.)
> 
> I wonder if we should run `insert_hooked_blocks_into_rest_response` from inside the Posts controller instead, in order to insert hooked blocks into every CPT that has `show_in_rest` set to `true` 🤔

Reply:
How about the following?
```suggestion
 * @since 6.6.0 Support `wp_navigation` post type.
 * @since 6.8.0 Support other post types.
```

Reply:
Sounds good. (The only thing we'll lose that way is the information that prior to 6.8.0, it was only `wp_navigation` post types that were supported, but maybe that's not that important.)

Reply:
I'll apply the change, and will add a note in parentheses that it was previously just `wp_navigation`.

```suggestion
 * @since 6.6.0
 * @since 6.8.0 Support other post types. (Previously, it was limited to `wp_navigation` only.)
```